code/modules/cargo/randomstock.dm

```diff
@@ -0,0 +1,1880 @@
+	//These are spawned inside the cargo warehouse, in crates and on tables.
+	var/spawns = TOTAL_STOCK
+	while (spawns > 0)
+		var/atom/spawnloc = get_location()
+		var/spawntype = get_spawntype()
+		spawn_stock(spawntype, spawnloc, src)
+		spawns--
+
+	handle_infestation()
+	shuffle_items()
+
+/datum/cargospawner/proc/get_location()
+	var/cratespawn = 0
+	var/obj/structure/closet/crate/emptiest
+	if (prob(70))//We'll usually put items in crates
+		var/minweight = 99999999999//We will distribute items somewhat evenly among crates
+		//by selecting the least-filled one for each spawn
+
+		for (var/obj/structure/closet/crate/C in containers)
+			if (C.stored_weight() < minweight && C.stored_weight() <= C.storage_capacity)
+				minweight = C.stored_weight()
+				emptiest = C
+				cratespawn = 1
+
+	if (cratespawn)
+		return emptiest
+	else
+		//If cratespawn is zero, we either failed to find a crate to spawn in, or didnt select
+		//crate spawning with the random check.
+		var/turf/clearest
+		var/min_items = 999999999
+		for (var/B in tables)//As with crates, we attempt to distribute items evenly. Pick least-filled
+			var/TB = get_turf(B)
+			var/numitems = 0
+			for (var/obj/item in TB)
+				numitems++
+			if (numitems <= min_items)
+				clearest = TB
+				min_items = numitems
+
+
+		if (!clearest)//This should never happen
+			clearest = get_turf(pick(tables))
+
+		return clearest
+
+
+/datum/cargospawner/proc/get_spawntype()
+	var/list/spawntypes = list("1" = STOCK_RARE_PROB, "2" = STOCK_UNCOMMON_PROB, "3" = (100 - (STOCK_RARE_PROB + STOCK_UNCOMMON_PROB)))
+	var/stocktype = pickweight(spawntypes)
+	switch (stocktype)
+		if ("1")
+			return pickweight(random_stock_rare)
+		if ("2")
+			return pickweight(random_stock_uncommon)
+		if ("3")
+			return pickweight(random_stock_common)
+
+//Minor and moderate mobs are checked per crate
+#define INFEST_PROB_MINOR	6
+#define INFEST_PROB_MODERATE	3
+
+#define INFEST_PROB_SEVERE	3//Severe is once per round, not per crate
+
+/datum/cargospawner/proc/handle_infestation()
+	for (var/obj/O in containers)
+		if (prob(INFEST_PROB_MINOR))
+	//No admin message for the minor mobs, because they are friendly and harmless
+			var/ctype = pickweight(infest_mobs_minor)
+			new ctype(O)
+		else if	 (prob(INFEST_PROB_MODERATE))
+			var/ctype = pickweight(infest_mobs_moderate)
+			new ctype(O)
+			msg_admin_attack("Common cargo warehouse critter [ctype] spawned inside [O.name] coords (<a href='?_src_=holder;adminplayerobservecoodjump=1;X=[O.x];Y=[O.y];Z=[O.z]'>JMP</a>)")
+
+
+	//This is checked only once per round. ~3% chance to spawn a scary monster infesting the warehouse
+	if (prob(INFEST_PROB_SEVERE))
+		//Find a tile to spawn the thing
+		var/list/turfs = list()
+		var/turf/T
+		for (var/turf/t in warehouse)
+			T = t//Failsafe incase none are clear
+			if (turf_clear(T))
+				turfs |= t
+
+		if (turfs.len)
+			T = pick(turfs)
+
+		var/ctype = pickweight(infest_mobs_severe)
+		if (ctype != "cratey")
```

Comment:
Would be cleaner to check if ctype is (as opposed to is not) cratey so other special mobs can be added easier with a else if later on